### PR TITLE
Signup: Add pure versions of steps/flows configs

### PR DESCRIPTION
**Pull request by jsnmoon**
While working on #25387, I discovered that `signup/config/flows` and `signup/config/steps` introduce a DOM API dependency due to imports such as `lib/signup/step-actions` and `lib/user`. 

In order to facilitate handling of the signup data flow via Redux, it is necessary to create pure, side-effects-free versions of these configuration files.

To this end, I have moved the flows and steps definitions into `signup/config/flows-pure` and `signup/config/steps-pure`, respectively. These pure counterparts export their configuration values without DOM API dependent values or functions. `signup/config/flows` and `signup/config/steps` have been kept backwards compatible for when such values or functions remain necessary.

In summary:
1. This PR creates `signup/config/flows-pure` and `signup/config/steps-pure` configuration files. These files replace their DOM dependent functions with noop's.
2. This PR preserves backward compatibliity for `signup/config/flows` and `signup/config/steps`.
3. Signup's dependency store, progress store, and flow controller have been refactored to use `signup/config/steps-pure`.

# Testing Instructions
1. Spin up this branch locally.
2. Go through the signup flow at [`/start`](http://calypso.localhost:3000/start/domains).
3. Ensure no behavioral changes have been introduced.

### CODE REVIEW

**Comment by jsnmoon** [comment]
Tests are not passing, marking as In Progress in order to investigate.

**Comment by jsnmoon** [comment]
Good questions - let me provide a little more context.

> What are the DOM-dependent imports? (@dmsnell)

For `signup/config/flows`, importing `lib/user` makes it DOM dependent due to the `lib/wp` import contained within `lib/user/user`. The user library is used here to check for the current session's authentication state.

Likewise for `signup/config/steps`, importing `lib/signup/step-actions` makes it DOM dependent as `step-actions` imports both `lib/user` and `lib/wp`. `lib/user` is used for both authentication and fetching user data; `lib/wp` is invoked directly to create new user accounts, validate new users, and create new sites using `wpcom.undocumented()` functions.

> It would be really helpful to understand what exactly is breaking down and motivating this change. (@dmsnell)

Basically, I want to be able to use steps and flows data within a reducer. However, to do so without a change like this would mean having to mock out `lib/user` and `lib/signup/step-actions` in every single reducer unit tests. I discovered this issue while working on reduxifying signup progress handling (#25387) where importing either configuration files in a single reducer would break all reducer unit tests.

> I'm also curious if there aren't ways that don't involve -pure.js file naming. Do we have specific things that are DOM dependent that could be extracted to remain clean otherwise? (@dmsnell)

The original idea I'd been toying with was to store steps and flows definitions inside respective JSON files (`steps.json`, `flows.json`) which could be enhanced with respective callback functions inside the actual modules (`steps.js`, `flows.js`). Perhaps this would be easier to understand than the `module-pure.js` nomenclature?

**Comment by dmsnell** [comment]
Can you share a bit more about how wrapping the configs in `generate…()` is helping? I didn't see the imports in there

**Comment by jsnmoon** [comment]
> Can you share a bit more about how wrapping the configs in `generate…()` is helping? I didn't see the imports in there (@dmsnell)

By default, `signup/config/flows-pure` and `signup/config/steps-pure` modules will export the configuration values without any of the function bindings. This is done [here for the flows](https://github.com/Automattic/wp-calypso/pull/25606/files#diff-28b03eff6ef567c591024d6033cda11dR332) and [here for the steps](https://github.com/Automattic/wp-calypso/pull/25606/files#diff-ae0bf020974ae29c034739f287afc5c5R343).

**Review by jsnajdr — APPROVED** [review]
The changes look good and signup still works.

It's likely that after the Signup Redux work is finished, we can merge the "pure" and "impure" files into one again? `lib/user` will be replaced by its Redux counterpart, `wpcom` calls will be replaced by Redux action dispatches...

Thanks @jsnmoon for working on the Flux removal!

File: client/lib/signup/dependency-store.js [hunk 1]
```diff
@@ -31,8 +31,8 @@ const SignupDependencyStore = {
 };
 
 function assertValidDependencies( action ) {
-	const providesDependencies = steps[ action.data.stepName ].providesDependencies || [],
-		extraDependencies = difference( keys( action.providedDependencies ), providesDependencies );
+	const providesDependencies = get( steps, [ action.data.stepName, 'providesDependencies' ], [] );
```
Comment (blowery) [inline client/lib/signup/dependency-store.js]: this has slightly different semantics than the old version. If `providesDependencies` exists, but is null, it will no longer default to `[]`.  See http://jsbin.com/wixexidito/edit?js,console

Comment (blowery) [reply client/lib/signup/dependency-store.js]: that may be ok. It just depends on what providesDependencies is when it's empty.
